File: Codes from the AI-assistants/[Experimental] regex-redux/Basic Prompt - Normal Mode/RegexRedux-claude-basic-normal.js
#!/usr/bin/env node
/**
 * Regex Redux Algorithm Implementation in JavaScript
 * Uses native Node.js modules for maximum performance
 */

const fs = require('fs');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');
const os = require('os');

// Worker function for parallel pattern matching
if (!isMainThread) {
    const { sequence, pattern } = workerData;
    const regex = new RegExp(pattern, 'gi');
    const matches = sequence.match(regex);
    parentPort.postMessage(matches ? matches.length : 0);
    process.exit(0);
}

async function regexReduxSolution() {
    // Read input from stdin
    let input = '';
    
    return new Promise((resolve) => {
        process.stdin.setEncoding('utf8');
        
        process.stdin.on('data', (chunk) => {
            input += chunk;
        });
        
        process.stdin.on('end', async () => {
            const originalLength = input.length;
            
            // Step 1: Remove sequence descriptions and newlines
            const sequence = input.replace(/>.*\n|\n/g, '');
            const cleanedLength = sequence.length;
            
            // Step 2: Define patterns to count
            const patterns = [
                'agggtaaa|tttaccct',
                '[cgt]gggtaaa|tttaccc[acg]',
                'a[act]ggtaaa|tttacc[agt]t',
                'ag[act]gtaaa|tttac[agt]ct',
                'agg[act]taaa|ttta[agt]cct',
                'aggg[acg]aaa|ttt[cgt]ccct',
                'agggt[cgt]aa|tt[acg]accct',
                'agggta[cgt]a|t[acg]taccct',
                'agggtaa[cgt]|[acg]ttaccct'
            ];
            
            // Step 3: Count patterns using Worker threads for parallel processing
            const counts = await countPatternsParallel(sequence, patterns);
            
            // Step 4: Apply replacements
            const replacements = [
                [/tHa[Nt]/gi, '<4>'],
                [/aND|caN|Ha[DS]|WaS/gi, '<3>'],
                [/a[NSt]|BY/gi, '<2>'],
                [/<[^>]*>/g, '|'],
                [/\|[^|][^|]*\|/g, '-']
            ];
            
            let resultSequence = sequence;
            for (const [pattern, replacement] of replacements) {
                resultSequence = resultSequence.replace(pattern, replacement);
            }
            
            const finalLength = resultSequence.length;
            
            // Output results
            patterns.forEach((pattern, i) => {
                console.log(`${pattern} ${counts[i]}`);
            });
            
            console.log(`\n${originalLength}`);
            console.log(`${cleanedLength}`);
            console.log(`${finalLength}`);
            
            resolve();
        });
    });
}

async function countPatternsParallel(sequence, patterns) {
    const numWorkers = Math.min(patterns.length, os.cpus().length);
    const promises = [];
    
    for (let i = 0; i < patterns.length; i++) {
        const promise = new Promise((resolve, reject) => {
            const worker = new Worker(__filename, {
                workerData: { sequence, pattern: patterns[i] }
            });
            
            worker.on('message', resolve);
            worker.on('error', reject);
        });
        promises.push(promise);
    }
    
    return Promise.all(promises);
}

// Optimized synchronous version
function optimizedRegexRedux() {
    const input = fs.readFileSync(0, 'utf8');
    const originalLength = input.length;
    
    // Pre-compile cleanup regex
    const cleanupRegex = />.*\n|\n/g;
    const sequence = input.replace(cleanupRegex, '');
    const cleanedLength = sequence.length;
    
    // Pre-compile counting patterns
    const countPatterns = [
        /agggtaaa|tttaccct/gi,
        /[cgt]gggtaaa|tttaccc[acg]/gi,
        /a[act]ggtaaa|tttacc[agt]t/gi,
        /ag[act]gtaaa|tttac[agt]ct/gi,
        /agg[act]taaa|ttta[agt]cct/gi,
        /aggg[acg]aaa|ttt[cgt]ccct/gi,
        /agggt[cgt]aa|tt[acg]accct/gi,
        /agggta[cgt]a|t[acg]taccct/gi,
        /agggtaa[cgt]|[acg]ttaccct/gi
    ];
    
    // Count patterns efficiently
    const counts = countPatterns.map(pattern => {
        const matches = sequence.match(pattern);
        return matches ? matches.length : 0;
    });
    
    // Pre-compile replacement patterns
    const replacements = [
        [/tHa[Nt]/gi, '<4>'],
        [/aND|caN|Ha[DS]|WaS/gi, '<3>'],
        [/a[NSt]|BY/gi, '<2>'],
        [/<[^>]*>/g, '|'],
        [/\|[^|][^|]*\|/g, '-']
    ];
    
    // Apply replacements
    let resultSequence = sequence;
    replacements.forEach(([pattern, replacement]) => {
        resultSequence = resultSequence.replace(pattern, replacement);
    });
    
    const finalLength = resultSequence.length;
    
    // Output results
    const patternStrings = [
        'agggtaaa|tttaccct',
        '[cgt]gggtaaa|tttaccc[acg]',
        'a[act]ggtaaa|tttacc[agt]t',
        'ag[act]gtaaa|tttac[agt]ct',
        'agg[act]taaa|ttta[agt]cct',
        'aggg[acg]aaa|ttt[cgt]ccct',
        'agggt[cgt]aa|tt[acg]accct',
        'agggta[cgt]a|t[acg]taccct',
        'agggtaa[cgt]|[acg]ttaccct'
    ];
    
    patternStrings.forEach((pattern, i) => {
        console.log(`${pattern} ${counts[i]}`);
    });
    
    console.log(`\n${originalLength}`);
    console.log(`${cleanedLength}`);
    console.log(`${finalLength}`);
}

// Stream-based version for large files
function streamBasedRegexRedux() {
    const readline = require('readline');
    
    const rl = readline.createInterface({
        input: process.stdin,
        crlfDelay: Infinity
    });
    
    let sequence = '';
    let originalLength = 0;
    
    rl.on('line', (line) => {
        originalLength += line.length + 1; // +1 for newline
        if (!line.startsWith('>')) {
            sequence += line;
        }
    });
    
    rl.on('close', () => {
        originalLength -= 1; // Adjust for last newline
        const cleanedLength = sequence.length;
        
        // Process patterns
        const patterns = [
            'agggtaaa|tttaccct',
            '[cgt]gggtaaa|tttaccc[acg]',
            'a[act]ggtaaa|tttacc[agt]t',
            'ag[act]gtaaa|tttac[agt]ct',
            'agg[act]taaa|ttta[agt]cct',
            'aggg[acg]aaa|ttt[cgt]ccct',
            'agggt[cgt]aa|tt[acg]accct',
            'agggta[cgt]a|t[acg]taccct',
            'agggtaa[cgt]|[acg]ttaccct'
        ];
        
        const counts = patterns.map(pattern => {
            const regex = new RegExp(pattern, 'gi');
            const matches = sequence.match(regex);
            return matches ? matches.length : 0;
        });
        
        // Apply replacements
        let result = sequence
            .replace(/tHa[Nt]/gi, '<4>')
            .replace(/aND|caN|Ha[DS]|WaS/gi, '<3>')
            .replace(/a[NSt]|BY/gi, '<2>')
            .replace(/<[^>]*>/g, '|')
            .replace(/\|[^|][^|]*\|/g, '-');
        
        const finalLength = result.length;
        
        // Output
        patterns.forEach((pattern, i) => {
            console.log(`${pattern} ${counts[i]}`);
        });
        
        console.log(`\n${originalLength}`);
        console.log(`${cleanedLength}`);
        console.log(`${finalLength}`);
    });
}

// Main execution
if (isMainThread) {
    // Choose implementation based on your needs:
    optimizedRegexRedux();        // Best for most cases
    // regexReduxSolution();      // Parallel processing version
    // streamBasedRegexRedux();   // Memory-efficient streaming version
}

File: Codes from the AI-assistants/[Experimental] regex-redux/Basic Prompt - Normal Mode/RegexRedux-claude-basic-normal.py
#!/usr/bin/env python3
"""
Regex Redux Algorithm Implementation
Uses native Python libraries for maximum performance
"""

import re
import sys
from concurrent.futures import ThreadPoolExecutor
import multiprocessing as mp

def regex_redux_solution():
    # Read input from stdin
    input_data = sys.stdin.read()
    
    # Store original length
    original_length = len(input_data)
    
    # Step 1: Remove sequence descriptions and newlines
    # This regex removes FASTA sequence descriptions (lines starting with '>') and newlines
    sequence = re.sub(r'>.*\n|\n', '', input_data)
    cleaned_length = len(sequence)
    
    # Step 2: Define regex patterns to count
    patterns = [
        r'agggtaaa|tttaccct',
        r'[cgt]gggtaaa|tttaccc[acg]',
        r'a[act]ggtaaa|tttacc[agt]t',
        r'ag[act]gtaaa|tttac[agt]ct',
        r'agg[act]taaa|ttta[agt]cct',
        r'aggg[acg]aaa|ttt[cgt]ccct',
        r'agggt[cgt]aa|tt[acg]accct',
        r'agggta[cgt]a|t[acg]taccct',
        r'agggtaa[cgt]|[acg]ttaccct'
    ]
    
    # Step 3: Count pattern matches using multiprocessing for performance
    def count_pattern(pattern):
        return len(re.findall(pattern, sequence, re.IGNORECASE))
    
    # Use ThreadPoolExecutor for I/O bound regex operations
    with ThreadPoolExecutor(max_workers=mp.cpu_count()) as executor:
        counts = list(executor.map(count_pattern, patterns))
    
    # Step 4: Define replacement patterns
    replacements = [
        (r'tHa[Nt]', '<4>'),
        (r'aND|caN|Ha[DS]|WaS', '<3>'),
        (r'a[NSt]|BY', '<2>'),
        (r'<[^>]*>', '|'),
        (r'\|[^|][^|]*\|', '-')
    ]
    
    # Step 5: Apply replacements sequentially
    result_sequence = sequence
    for pattern, replacement in replacements:
        result_sequence = re.sub(pattern, replacement, result_sequence, flags=re.IGNORECASE)
    
    final_length = len(result_sequence)
    
    # Output results
    for i, (pattern, count) in enumerate(zip(patterns, counts)):
        print(f"{pattern} {count}")
    
    print(f"\n{original_length}")
    print(f"{cleaned_length}")
    print(f"{final_length}")

def optimized_regex_redux():
    """
    Optimized version with pre-compiled patterns and memory-efficient processing
    """
    input_data = sys.stdin.read()
    original_length = len(input_data)
    
    # Pre-compile regex patterns for better performance
    cleanup_pattern = re.compile(r'>.*\n|\n')
    sequence = cleanup_pattern.sub('', input_data)
    cleaned_length = len(sequence)
    
    # Pre-compile counting patterns
    count_patterns = [
        re.compile(r'agggtaaa|tttaccct', re.IGNORECASE),
        re.compile(r'[cgt]gggtaaa|tttaccc[acg]', re.IGNORECASE),
        re.compile(r'a[act]ggtaaa|tttacc[agt]t', re.IGNORECASE),
        re.compile(r'ag[act]gtaaa|tttac[agt]ct', re.IGNORECASE),
        re.compile(r'agg[act]taaa|ttta[agt]cct', re.IGNORECASE),
        re.compile(r'aggg[acg]aaa|ttt[cgt]ccct', re.IGNORECASE),
        re.compile(r'agggt[cgt]aa|tt[acg]accct', re.IGNORECASE),
        re.compile(r'agggta[cgt]a|t[acg]taccct', re.IGNORECASE),
        re.compile(r'agggtaa[cgt]|[acg]ttaccct', re.IGNORECASE)
    ]
    
    # Count patterns using list comprehension for efficiency
    counts = [len(pattern.findall(sequence)) for pattern in count_patterns]
    
    # Pre-compile replacement patterns
    replacement_patterns = [
        (re.compile(r'tHa[Nt]', re.IGNORECASE), '<4>'),
        (re.compile(r'aND|caN|Ha[DS]|WaS', re.IGNORECASE), '<3>'),
        (re.compile(r'a[NSt]|BY', re.IGNORECASE), '<2>'),
        (re.compile(r'<[^>]*>'), '|'),
        (re.compile(r'\|[^|][^|]*\|'), '-')
    ]
    
    # Apply replacements with compiled patterns
    result_sequence = sequence
    for pattern, replacement in replacement_patterns:
        result_sequence = pattern.sub(replacement, result_sequence)
    
    final_length = len(result_sequence)
    
    # Output results
    pattern_strings = [
        'agggtaaa|tttaccct',
        '[cgt]gggtaaa|tttaccc[acg]',
        'a[act]ggtaaa|tttacc[agt]t',
        'ag[act]gtaaa|tttac[agt]ct',
        'agg[act]taaa|ttta[agt]cct',
        'aggg[acg]aaa|ttt[cgt]ccct',
        'agggt[cgt]aa|tt[acg]accct',
        'agggta[cgt]a|t[acg]taccct',
        'agggtaa[cgt]|[acg]ttaccct'
    ]
    
    for pattern_str, count in zip(pattern_strings, counts):
        print(f"{pattern_str} {count}")
    
    print(f"\n{original_length}")
    print(f"{cleaned_length}")
    print(f"{final_length}")

def memory_efficient_regex_redux():
    """
    Memory-efficient version for very large inputs
    """
    import mmap
    
    # For very large files, use memory mapping
    with open(0, 'rb') as f:
        with mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) as mmapped_file:
            input_data = mmapped_file.read().decode('utf-8')
    
    original_length = len(input_data)
    
    # Process in chunks if needed for extremely large datasets
    sequence = re.sub(r'>.*\n|\n', '', input_data)
    cleaned_length = len(sequence)
    
    # Use generator expressions for memory efficiency
    patterns = [
        r'agggtaaa|tttaccct',
        r'[cgt]gggtaaa|tttaccc[acg]',
        r'a[act]ggtaaa|tttacc[agt]t',
        r'ag[act]gtaaa|tttac[agt]ct',
        r'agg[act]taaa|ttta[agt]cct',
        r'aggg[acg]aaa|ttt[cgt]ccct',
        r'agggt[cgt]aa|tt[acg]accct',
        r'agggta[cgt]a|t[acg]taccct',
        r'agggtaa[cgt]|[acg]ttaccct'
    ]
    
    counts = []
    for pattern in patterns:
        count = len(re.findall(pattern, sequence, re.IGNORECASE))
        counts.append(count)
        print(f"{pattern} {count}")
    
    # Apply replacements
    replacements = [
        (r'tHa[Nt]', '<4>'),
        (r'aND|caN|Ha[DS]|WaS', '<3>'),
        (r'a[NSt]|BY', '<2>'),
        (r'<[^>]*>', '|'),
        (r'\|[^|][^|]*\|', '-')
    ]
    
    result_sequence = sequence
    for pattern, replacement in replacements:
        result_sequence = re.sub(pattern, replacement, result_sequence, re.IGNORECASE)
    
    final_length = len(result_sequence)
    
    print(f"\n{original_length}")
    print(f"{cleaned_length}")
    print(f"{final_length}")

if __name__ == "__main__":
    # Choose the implementation based on your needs:
    # optimized_regex_redux()        # Best for most cases
    # memory_efficient_regex_redux() # For very large inputs
    # regex_redux_solution()         # Basic implementation with multiprocessing
    
    optimized_regex_redux()